src/cpp/shared/combat/army.cpp

```diff
@@ -1343,83 +1345,100 @@ void SpecialAttackGraphics(army *attacker, army *target) {
-      (*(float *)((char *)&attacker->frameInfo.projectileStartOffset[i + 2] + 1) +
-        attacker->frameInfo.projDirAngle[i]) / 2.0 >= angleDeg;  ++i)
-      ;
-    if (attacker->frameInfo.numMissileDirs <= i)
-      spriteIdx = attacker->frameInfo.numMissileDirs - 1;
-    else
-      spriteIdx = i - 1;
-  } else {
-    if (yDiff <= 0)
-      spriteIdx = 0;
-    else
-      spriteIdx = attacker->frameInfo.numMissileDirs - 1;
-    angleDeg = (double)(yDiff <= 0 ? 90 : -90);
+  {
+    if (totXDiff) {
+      float slope = (double)-yDiff / (double)totXDiff;
+      angleDeg = atan(slope) * 180.0 / 3.14159;
+      int i;
+      for (i = 1;  attacker->frameInfo.numMissileDirs > i &&
+        (*(float *)((char *)&attacker->frameInfo.projectileStartOffset[i + 2] + 1) +
+          attacker->frameInfo.projDirAngle[i]) / 2.0 >= angleDeg;  ++i)
+        ;
+      if (attacker->frameInfo.numMissileDirs <= i)
+        spriteIdx = attacker->frameInfo.numMissileDirs - 1;
+      else
+        spriteIdx = i - 1;
+    }
+    else {
+      if (yDiff <= 0)
+        spriteIdx = 0;
+      else
+        spriteIdx = attacker->frameInfo.numMissileDirs - 1;
+      angleDeg = (double)(yDiff <= 0 ? 90 : -90);
+    }
   }
 
   int attackDirectionAnimationIdx;
-  if (angleDeg <= 25.0) {
-    if (angleDeg <= -25.0) {
-      attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_DOWNWARDS;
-      attackDirectionAnimationIdx = 2;
+  {
+    if (angleDeg <= 25.0) {
+      if (angleDeg <= -25.0) {
+        attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_DOWNWARDS;
+        attackDirectionAnimationIdx = 2;
+      } else {
+        attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_FORWARDS;
+        attackDirectionAnimationIdx = 1;
+      }
     } else {
-      attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_FORWARDS;
-      attackDirectionAnimationIdx = 1;
+      attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_UPWARDS;
+      attackDirectionAnimationIdx = 0;
     }
-  } else {
-    attacker->animationType = ANIMATION_TYPE_RANGED_ATTACK_UPWARDS;
-    attackDirectionAnimationIdx = 0;
-  }
-  for (attacker->animationFrame = 0;
-    attacker->frameInfo.animationLengths[attacker->animationType] > attacker->animationFrame;
-    ++attacker->animationFrame) {
-    if (attacker->frameInfo.animationLengths[attacker->animationType] - 1 == attacker->animationFrame)
-      gpCombatManager->DrawFrame(0, 1, 0, 0, 75, 1, 1);
-    else
-      gpCombatManager->DrawFrame(1, 1, 0, 0, 75, 1, 1);
-    glTimers = (signed __int64)((double)KBTickCount()
-      + (double)attacker->frameInfo.shootingTime
-      * gfCombatSpeedMod[giCombatSpeed]
-      / (double)attacker->frameInfo.animationLengths[attacker->animationType]);
-  }
-  attacker->animationFrame = attacker->frameInfo.animationLengths[attacker->animationType] - 1;
-  int projIconWidth = 100;
-  int projIconHeight = 100;
-  int v35 = 31;
+  }
+
+  DrawShootingAnimation: {
```

Comment:
A goto label without goto? :thinking:

Reply:
Yep, exactly. It's a label for this block of code. I'd label the other bare-braced blocks if I had a good name for them.

Reply:
Doesn't this _ask_ to be moved into a separate function then? Same for scopes I mentioned above

Reply:
This one looks like it can be pretty manageably. The ones above less so because they have multiple outputs. Identifying a block like this and labeling it is pretty much always a win. It takes more thought to decide whether splitting something into a function is a good idea because there are more tradeoffs.